### PR TITLE
Initial modifications to get enums working with postgres schemas

### PR DESCRIPTION
**Pull request by faceleg**
First off, thanks for all the hard work, I'm loving Sequelize. 

I am using Postgres and want to use a non-standard schema for some models. I use ENUM for my model `status` fields, which has lead to problems for me when combined with the non-standard schema.

This PR aims to make Sequelize work perfectly with non standard Postgres schemas.

Please can you glance through the commit and let me know if you're happy with the direction I'm heading. Once I have it working I'll look at extracting the duplication in to a helper function or two.

### CODE REVIEW

File: lib/dialects/postgres/query-generator.js [hunk 1]
```diff
@@ -758,7 +753,10 @@ var QueryGenerator = {
       values = dataType.toString().match(/^ENUM\(.+\)/)[0];
     }
 
-    var sql = 'CREATE TYPE ' + enumName + ' AS ' + values + ';';
+    var schema = tableName.schema || options.schema || 'public';
+    var delimiter = tableName.delimiter || options.delimiter || '.';
+
+    var sql = 'CREATE TYPE ' + this.quoteIdentifier(schema) + delimiter + enumName + ' AS ' + values + ';';
```
Comment (janmeier) [inline lib/dialects/postgres/query-generator.js]: Why this change - the name creation is intentionally moved to a function, which handles schema? Isn't the end output the same?

Comment (faceleg) [inline lib/dialects/postgres/query-generator.js]: It was causing this portion (L746) to fail:

```JS
var query = 'SELECT t.typname enum_name, array_agg(e.enumlabel) enum_value FROM pg_type t ' +
       'JOIN pg_enum e ON t.oid = e.enumtypid ' +
       'JOIN pg_catalog.pg_namespace n ON n.oid = t.typnamespace ' +
       "WHERE n.nspname = '" + schema + "'" + enumName + ' GROUP BY 1';
```

Comment (janmeier) [inline lib/dialects/postgres/query-generator.js]: Then the change should be the other way I think - so that enumName is generated manually here. `pgEnumName` is used name places, otherwise we'd have to duplicate the logic for adding schema in front of enum name to all those functiond

Comment (faceleg) [inline lib/dialects/postgres/query-generator.js]: I was just typing the same thing :dancer: 

Too many late late nights in a row

**Comment by faceleg** [comment]
@janmeier another go

**Comment by faceleg** [comment]
@janmeier ready for your scrutiny

**Comment by faceleg** [comment]
@janmeier bump

**Comment by janmeier** [comment]
One failing test, otherwise LGTM.

If you could fix the test (just add a check for options), add a changelog entry and squash to a single descriptive commit this is good to go

**Comment by faceleg** [comment]
Is it expected that pgListEnums would get called with no tableName?

**Comment by janmeier** [comment]
Yes, it can also be used to list all existing enums, not just those in a single table

**Comment by faceleg** [comment]
Please let me know if this is suitable

**Comment by faceleg** [comment]
Bump

**Comment by faceleg** [comment]
Bumping again :dancer:

**Comment by faceleg** [comment]
@mickhansen if @janmeier is super busy, would you please be able to approve this?

File: lib/dialects/postgres/query-generator.js [hunk 2]
```diff
@@ -725,16 +725,23 @@ var QueryGenerator = {
 
     var enumName = '"enum_' + (_.isPlainObject(tableName) ? tableName.table : tableName) + '_' + attr + '"';
 
-    if (options.schema !== false && tableName && tableName.schema) {
-      enumName = '"' + tableName.schema + '"' + tableName.delimiter + enumName;
+    if (options.schema !== false && (tableName.schema || options.schema)) {
+      var schema = tableName.schema || options.schema || 'public';
+      var delimiter = tableName.delimiter || options.delimiter || '.';
```
Comment (mickhansen) [inline lib/dialects/postgres/query-generator.js]: Please use pre-comma per the general style,

```js
var schema = ...

```

File: test/unit/sql/enum.test.js [hunk 3]
```diff
@@ -24,6 +25,22 @@ describe(Support.getTestDialectTeaser('SQL'), function() {
         }
       });
 
+      describe('pgEnumName', function() {
+        it('does not add schema when options: { schema: false }', function() {
+          expect(sql.pgEnumName(PublicUser.getTableName(), 'mood', { schema: false }))
+            .to.equal('"enum_users_mood"');
+          expect(sql.pgEnumName(FooUser.getTableName(), 'theirMood', { schema: false }))
+            .to.equal('"enum_users_theirMood"');
```
Comment (mickhansen) [inline test/unit/sql/enum.test.js]: When would a user specifically use `schema: false`? Would it not normally just be unset?

File: test/unit/sql/enum.test.js [hunk 4]
```diff
@@ -45,9 +62,9 @@ describe(Support.getTestDialectTeaser('SQL'), function() {
           });
         });
 
-        it('uses the default schema if options given', function () {
+        it('uses the default schema if no options given', function () {
             expectsql(sql.pgListEnums(), {
-            postgres: 'SELECT t.typname enum_name, array_agg(e.enumlabel) enum_value FROM pg_type t JOIN pg_enum e ON t.oid = e.enumtypid JOIN pg_catalog.pg_namespace n ON n.oid = t.typnamespace WHERE n.nspname = \'public\' GROUP BY 1'
+              postgres: 'SELECT t.typname enum_name, array_agg(e.enumlabel) enum_value FROM pg_type t JOIN pg_enum e ON t.oid = e.enumtypid JOIN pg_catalog.pg_namespace n ON n.oid = t.typnamespace WHERE n.nspname = \'public\' GROUP BY 1'
           });
```
Comment (mickhansen) [inline test/unit/sql/enum.test.js]: I see you fixed the indentation here, looks like it was `expectsql` that needed to be moved back tho :)

File: lib/dialects/postgres/connection-manager.js [hunk 5]
```diff
@@ -119,6 +119,7 @@ ConnectionManager.prototype.disconnect = function(connection) {
   return new Promise(function (resolve, reject) {
     connection.end();
     resolve();
+    return null;
```
Comment (mickhansen) [inline lib/dialects/postgres/connection-manager.js]: Any specific reason for this?

File: lib/model.js [hunk 6]
```diff
@@ -2453,6 +2453,7 @@ Model.prototype.update = function(values, options) {
         }
       });
     }
+    return null;
```
Comment (mickhansen) [inline lib/model.js]: same here, is it just for specificity?

File: lib/dialects/postgres/query-generator.js [hunk 7]
```diff
@@ -725,16 +725,23 @@ var QueryGenerator = {
...
-  pgListEnums: function(tableName, attrName) {
-    var enumName = ''
-      , schema = tableName && tableName.schema ? tableName.schema : 'public';
+  pgListEnums: function(tableName, attrName, options) {
+    var enumName = '';
+    options =  options || {};
+    var schema = 'public';
+    if (tableName) {
+      schema = tableName.schema || options.schema || 'public';
+    }
```
Comment (mickhansen) [inline lib/dialects/postgres/query-generator.js]: Would `tableName` not always be set? But it might not always be an object.

File: lib/dialects/postgres/query-generator.js (hunk 2 above)
Comment (mickhansen) [inline lib/dialects/postgres/query-generator.js]: This code handles having schema/delimter be part of the table name and be part of the options.
I imagine we'll need this multiple places since a lot of queryInterface (migration) calls should support both, ideally we could provide some helper that takes a tableName and an options and returns schema, delimiter and tableName.

File: test/unit/sql/enum.test.js (hunk 3 above)
Comment (faceleg) [inline test/unit/sql/enum.test.js]: It's done here: https://github.com/sequelize/sequelize/blob/master/lib/dialects/postgres/query-generator.js#L737

File: lib/model.js (hunk 6 above)
Comment (faceleg) [inline lib/model.js]: http://bluebirdjs.com/docs/warning-explanations.html#warning-a-promise-was-created-in-a-handler-but-none-were-returned-from-it

File: lib/dialects/postgres/connection-manager.js (hunk 5 above)
Comment (faceleg) [inline lib/dialects/postgres/connection-manager.js]: Attempting to squash warnings - http://bluebirdjs.com/docs/warning-explanations.html#warning-a-promise-was-created-in-a-handler-but-none-were-returned-from-it

Comment (mickhansen) [inline lib/dialects/postgres/connection-manager.js]: Ah, makes sense for `then`, sure it's required for `new Promise`?

Comment (faceleg) [inline lib/dialects/postgres/connection-manager.js]: I have no idea - I'm getting this warning for this line in my CI tests, this was my attempt to remove the warning (didn't work).

File: test/unit/sql/enum.test.js (hunk 4 above)
Comment (faceleg) [inline test/unit/sql/enum.test.js]: Fixed

File: lib/dialects/postgres/query-generator.js [hunk 8]
```diff
@@ -720,36 +720,50 @@ var QueryGenerator = {
...
     options = options || {};
+    tableName = tableName || {};
+    var tableDetails = {
+      schema: tableName.schema || options.schema || 'public',
+      tableName: _.isPlainObject(tableName) ? tableName.table : tableName,
+      delimiter: tableName.delimiter || options.delimiter || '.'
+    };
+
+    return tableDetails;
+  },
```
Comment (mickhansen) [inline lib/dialects/postgres/query-generator.js]: Should work for mssql aswell, might want to move the method to abstractQueryGenerator

Comment (mickhansen) [inline lib/dialects/postgres/query-generator.js]: No real need for the intermediary variable either (just return the object).

Comment (faceleg) [inline lib/dialects/postgres/query-generator.js]: Done. When you've approved this PR in principle I will write some tests for extractTableDetails

File: lib/dialects/postgres/query-generator.js [hunk 9]
```diff
@@ -722,34 +722,36 @@ var QueryGenerator = {
 
   pgEnumName: function (tableName, attr, options) {
     options = options || {};
+    var tableDetails = this.extractTableDetails(tableName, options || {})
```
Comment (mickhansen) [inline lib/dialects/postgres/query-generator.js]: options has already been defaulted, no need to do it again

**Comment by mickhansen** [comment]
Linter errors!

**Comment by faceleg** [comment]
Never ending pull request

**Comment by mickhansen** [comment]
Well a `make jshint` woulda helped you on that one ;)

**Comment by faceleg** [comment]
Yep my bad :D ill have to deal with it in 12 hours it's sleep time

**Comment by faceleg** [comment]
@mickhansen updated

**Comment by mickhansen** [comment]
postgres breaks: https://travis-ci.org/sequelize/sequelize/jobs/90589723

**Comment by faceleg** [comment]
:smile: :palm_tree:

**Comment by faceleg** [comment]
Huh, they pass on my machine. Looking into it

**Comment by faceleg** [comment]
Turns out I was running **dramatic pause** *the wrong tests*

**Comment by mickhansen** [comment]
https://travis-ci.org/sequelize/sequelize/jobs/90688045
Just a single expectation that's off.

**Comment by faceleg** [comment]
Oh look, I did fix it! (went to bed after last commit)

**Comment by faceleg** [comment]
Ahem, I mean, it's ready to merge.

**Comment by mickhansen** [comment]
Thank you for the good work and the patience @faceleg :)

**Comment by faceleg** [comment]
Pretty sure you were the patient one haha.

Thanks for bearing with me.
